keps/sig-network/3705-cloud-node-ips/README.md

```diff
@@ -0,0 +1,638 @@
+  Kubernetes cluster. This means that the default behavior when
+  `--node-ip` is not specified will remain the same, and the behavior
+  of any currently-allowed `--node-ip` value will remain the same. New
+  behavior will only be triggered by `--node-ip` values that would
+  have been rejected by kubelet in older clusters.
+
+- Adding the ability for nodes in clusters with cloud providers to use
+  node IPs that are not already known to the cloud provider. (In
+  particular, this implies that we will continue to not support
+  dual-stack nodes in clouds that do not themselves support
+  dual-stack.)
+
+- Improving the behavior of autodetecting the node IP in any other
+  ways. (Eg, being able to pass a CIDR rather than an IP.) There was
+  some discussion of ways we might do this in the abandoned [KEP-1664]
+  (which this KEP is a spiritual successor to), but if we want that,
+  it can be done later, independently of these changes. (These changes
+  were mostly desired in the context of non-cloud-provider clusters
+  anyway.)
+
+[KEP-1664]: https://github.com/kubernetes/enhancements/issues/1664
+
+## Proposal
+
+### Risks and Mitigations
+
+As the intention is to not change the user-visible behavior except in
+clusters where administrators explicitly make use of the new
+functionality, there should be no risk of breaking existing clusters,
+nor of surprising administrators by suddenly exposing node services on
+unexpected IPs.
+
+## Design Details
+
+### Current behavior
+
+Currently, when `--cloud-provider` is passed to kubelet, kubelet
+expects `--node-ip` to be either unset, or a single IP address. (If it
+is unset, that is equivalent to passing `--node-ip 0.0.0.0`, which
+means "autodetect an IPv4 address, or if there are no usable IPv4
+addresses, autodetect an IPv6 address".)
+
+If `--cloud-provider` and `--node-ip` are both specified (and
+`--node-ip` is not "`0.0.0.0`" or "`::`"), then kubelet will add an
+annotation to the node, `alpha.kubernetes.io/provided-node-ip`. Cloud
+providers expect this annotation to conform to the current expected
+`--node-ip` syntax (ie, a single value); if it does not, then they
+will log an error and, not remove the
+`node.cloudprovider.kubernetes.io/uninitialized` taint from the node,
+causing the node to remain unusable until kubelet is restarted with a
+valid (or absent) `--node-ip`.
+
+When `--cloud-provider` is not passed, the `--node-ip` value can also
+be a comma-separated pair of dual-stack IP addresses. However, unlike
+in the single-stack case, the IPs in the dual-stack case are not
+currently allowed to be "unspecified" IPs (ie `0.0.0.0` or `::`); you
+can only make a (non-cloud) node be dual-stack if you explicitly
+specify both IPs that you want it to use.
+
+### Changes to `--node-ip`
+
+The most obvious required change is that we need to allow dual-stack
+`--node-ip` values in clusters using external cloud providers (but
+_not_ in clusters using legacy cloud providers).
```

Comment:
my memory is fuzzy, how do you detect this?

Reply:
Currently kubelet has a `--cloud-provider` flag, which is either unset, `"external"`, or the name of one of the legacy providers. https://github.com/kubernetes/kubernetes/pull/110066 adds a config option to replace the `--cloud-provider` flag (with a boolean "has (external) cloud provider" / "doesn't have (any) cloud provider") when the legacy providers go away.